File: python/day1/main.py
f = open("input.txt")
lines = f.readlines()

# Part 1
def findFirstDigit(line: str) -> int:
    for c in line:
        if c.isnumeric():
            return int(c)

def findLastDigit(line: str) -> int:
    for c in reversed(line):
        if c.isnumeric():
            return int(c)

calibration_sum = 0

for line in lines:
    first = findFirstDigit(line)
    last = findLastDigit(line)
    calibration_sum += 10 * first + last

print(" sum = " + str(calibration_sum))


# Part 2
str_digits = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

def findRealFirstDigit(line: str) -> int:
    first_occurence = len(line)
    found_str_digit = None
    for i, digit in enumerate(str_digits):
        index = line.find(digit)

        if index > -1 and first_occurence > index:
            first_occurence = index
            found_str_digit = i + 1

    for i, c in enumerate(line):
        if c.isnumeric():
            if first_occurence < i:
                return found_str_digit
            else:
                return int(c)

def findRealLastDigit(line: str) -> int:
    first_occurence = 0
    found_str_digit = None

    for i, digit in enumerate(str_digits):
        try:
            index = line.rindex(digit) + 1
            if first_occurence < index:
                first_occurence = index
                found_str_digit = i + 1
        except:
            continue

    for i, c in enumerate(reversed(line)):
        if c.isnumeric():
            if first_occurence > len(line) - i:
                return found_str_digit
            else:
                return int(c)

calibration_sum = 0
for i, line in enumerate(lines):
    first = findRealFirstDigit(line)
    last = findRealLastDigit(line)
    calibration_sum += 10 * first + last

print(calibration_sum)


        

File: rust/day1/src/main.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn main() {
    // Part 1
    let mut result: u32 = 0;
    if let Ok(lines) = read_lines("src/input.txt") {
        for line_opt in lines {
            if let Ok(line) = line_opt {
                let line_value: u32 = get_line_value(&line);
                result += line_value;
            }
        }
    }

    println!("part1: {}", result);

    // Part 2
    let mut result: u32 = 0;
    if let Ok(lines) = read_lines("src/input.txt") {
        for line_opt in lines {
            if let Ok(line) = line_opt {
                let line_value: u32 = get_real_line_value(&line);
                result += line_value;
            }
        }
    }

    println!("part2: {}", result);
}

fn get_real_line_value(line: &str) -> u32 {
    let mut all_numbers: Vec<u32> = Vec::new();
    let digit_words: [&str; 9] = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];
    for (i, c) in line.chars().enumerate() {
        if c.is_numeric() {
            if let Some(number) = c.to_digit(10) {
                all_numbers.push(number);
            }
            continue;
        }
        for (word_index, word) in digit_words.iter().enumerate() {
            if line[i..].starts_with(word) {
                if let Ok(real_number) = u32::try_from(word_index) {
                    all_numbers.push(real_number + 1);
                }
            }
        }
    }

    if let Some(first) = all_numbers.first() {
        if let Some(last) = all_numbers.last() {
            return first * 10 + last;
        }
    }

    0
}

fn get_line_value(line: &str) -> u32 {
    let mut all_numbers: Vec<u32> = Vec::new();
    for c in line.chars() {
        if c.is_numeric() {
            if let Some(number) = c.to_digit(10) {
                all_numbers.push(number)
            }
        }
    }

    if let Some(first) = all_numbers.first() {
        if let Some(last) = all_numbers.last() {
            return first * 10 + last;
        }
    }

    0
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>> where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}